File: study/lua/src/app/fight/views/BillBoard.lua
local BillBoard =
    class(
    "BillBoard",
    function()
        return cc.Node:create()
    end
)

local Color = require("app.utils.Color")
local UIHelper = require("yoka.utils.UIHelper")
local HeroSkillEffect = require("app.config.hero_skill_effect")
local Path = require("app.utils.Path")
local FightConfig = require("app.fight.Config")
local EffectActor = require("app.fight.views.EffectActor")
local SchedulerHelper = require("app.utils.SchedulerHelper")

function BillBoard:ctor(name, quality, rankLevel, isPlayer, officelLevel, showMark, camp, maxHp, trueQuality)
    self:setCascadeOpacityEnabled(true)
    self:setCascadeColorEnabled(true)

    --血条黑低
    local panelHpBase = cc.Sprite:create(Path.getBattleRes("img_battle_hpbg"))
    self:addChild(panelHpBase)
    panelHpBase:setPosition(cc.p(0, 5))
    self._panelHpBase = panelHpBase

    --血条红底
    -- self._hpShadow = cc.Sprite:create(Path.getBattleRes('img_battle_hpshadow'))
    -- panelHpBase:addChild(self._hpShadow)
    -- self._hpShadow:setAnchorPoint(cc.p(0, 0))
    -- self._hpShadow:setPosition(cc.p(1, 2))
    self._hpShadow = ccui.LoadingBar:create(Path.getBattleRes("img_battle_hpshadow"), 100)
    panelHpBase:addChild(self._hpShadow)
    self._hpShadow:setAnchorPoint(cc.p(0, 0))
    self._hpShadow:setPosition(cc.p(1, 2))

    --白色护盾
    -- self._protect = ccui.LoadingBar:create(Path.getBattleRes("img_battle_es"), 0)
    -- panelHpBase:addChild(self._protect)
    -- self._protect:setAnchorPoint(cc.p(0, 0))
    -- self._protect:setPosition(cc.p(1, 2))
    -- self._protect:setPercent(0)

    self._protect = {}

    --绿色血条
    self._hp = ccui.LoadingBar:create(Path.getBattleRes("img_battle_hp"), 100)
    panelHpBase:addChild(self._hp)
    self._hp:setAnchorPoint(cc.p(0, 0))
    self._hp:setPosition(cc.p(1, 2))

    local angerBase = cc.Sprite:create(Path.getBattleRes("img_battle_bar_angebg"))
    self:addChild(angerBase)
    angerBase:setPosition(cc.p(0, -2))

    self._imageAnger = {}
    for i = 1, 4 do
        self._imageAnger[i] = cc.Sprite:create(Path.getBattleRes("img_battle_bar_ange0" .. i))
        self._imageAnger[i]:setPosition(cc.p(10 + (i - 1) * 16, 4))
        angerBase:addChild(self._imageAnger[i])
    end

    self._angerCountBG = cc.Sprite:create(Path.getBattleRes("img_battle_angenumbg"))
    self:addChild(self._angerCountBG)
    self._angerCountBG:setPosition(cc.p(50, 3))

    self._angerLabel = cc.Label:createWithCharMap(Path.getBattleFont("img_battle_angenum"), 20, 18, string.byte("+"))
    self._angerCountBG:addChild(self._angerLabel)
    self._angerLabel:setString("9")
    self._angerLabel:setAnchorPoint(cc.p(0, 0.5))
    self._angerLabel:setPosition(cc.p(20, 10))

    if rankLevel ~= 0 then
        if trueQuality == 7 and not isPlayer then -- 金将、
            name = name .. " " .. Lang.get("goldenhero_train_text") .. rankLevel
        else
            name = name .. "+" .. rankLevel
        end
    end
    local labelName = cc.Label:createWithTTF(name, Path.getCommonFont(), 22)
    self:addChild(labelName)
    labelName:setPosition(cc.p(0, 22))
    if isPlayer then
        labelName:setColor(Color.getOfficialColor(officelLevel))
        labelName:enableOutline(Color.getOfficialColorOutline(officelLevel), 2)
    else
        labelName:setColor(Color.getColor(quality))
        labelName:enableOutline(Color.getColorOutline(quality), 2)
    end

    if officelLevel and officelLevel ~= 0 and isPlayer then
        local officalInfo, officalLevel = G_UserData:getBase():getOfficialInfo(officelLevel)
        local labelOfficalName = cc.Label:createWithTTF("[" .. officalInfo.name .. "]", Path.getCommonFont(), 22)
        self:addChild(labelOfficalName)
        labelOfficalName:setPosition(cc.p(0, 45))
        labelOfficalName:setColor(Color.getOfficialColor(officalLevel))
        labelOfficalName:enableOutline(Color.getOfficialColorOutline(officalLevel), 2)
    end

    local markStartPosX = labelName:getContentSize().width / 2
    self:_createMarks(showMark, camp)

    self._buffIcons = {}
    for i = 1, 4 do
        local params = {}
        self._buffIcons[i] = UIHelper.createImage(params)
        self._buffIcons[i]:setAnchorPoint(cc.p(0, 0))
        self._buffIcons[i]:setPosition(cc.p(-40 + 20 * (i - 1), -28))
        self._buffIcons[i]:setScale(0.5)
        self:addChild(self._buffIcons[i])
        self._buffIcons[i]:setVisible(false)
    end

    self._maxHp = maxHp --最大血量
    self._hpPercent = 100 --目标百分比
    self._shadowPercent = 100 --背景红色的百分比
    self._angerEffect = nil
    self:setVisible(false)
end

function BillBoard:_createNewProtect()
    local protect = ccui.LoadingBar:create(Path.getBattleRes("img_battle_es"), 0)
    self._panelHpBase:addChild(protect)
    protect:setAnchorPoint(cc.p(0, 0))
    protect:setPosition(cc.p(1, 2))
    protect:setPercent(0)
    table.insert(self._protect, protect)
    protect:setLocalZOrder(#self._protect)
end

function BillBoard:_createMarks(showMark, camp)
    for i, v in pairs(showMark) do
        local effect = EffectActor.new(FightConfig.MARK[v])
        effect:setAction("effect", true)
        self:addChild(effect)
        -- effect:setScale(1.3)
        effect:setPositionY(-15)
        if camp == 2 then
            effect:setScaleX(-1)
        end
    end
end

--更新绿色的血条
function BillBoard:updateHP(hp, protect)
    local hpPercent = hp / self._maxHp * 100
    if hpPercent > 100 then
        hpPercent = 100
    end
    self._shadowPercent = hpPercent
    self._hp:setPercent(hpPercent)

    if protect ~= 0 then
        local protectCnt = math.ceil(protect / self._maxHp)
        if protectCnt ~= 0 then
            while not self._protect[protectCnt] do
                self:_createNewProtect()
            end
            for i = 1, protectCnt - 1 do
                self._protect[i]:setPercent(100)
            end
            if protectCnt < #self._protect then 
                for i = #self._protect, protectCnt + 1, -1  do 
                    self._protect[i]:removeFromParent()
                    table.remove(self._protect, i)
                end
            end
            local lastProtect = protect - (protectCnt - 1) * self._maxHp
            local protectPercent = lastProtect / self._maxHp * 100
            self._protect[#self._protect]:setPercent(protectPercent)
        end
    else 
        for i, v in pairs(self._protect) do 
            v:removeFromParent()
        end
        self._protect = {}
    end

    -- if protect == 0 then
    --     self._protect:setVisible(false)
    -- else
    --     self._protect:setVisible(true)
    -- end
    -- local percent = (hp + protect) / self._maxHp * 100
    -- if percent > 100 then
    --     self._shadowPercent = 100
    --     local hpPercent = hp / (hp + protect) * 100
    --     self._hp:setPercent(hpPercent)
    --     if protect == 0 then
    --         self._protect:setVisible(false)
    --     else
    --         self._protect:setVisible(true)
    --         self._protect:setPercent(100)
    --     end
    -- else
    --     local protectPercent = (protect + hp )/ self._maxHp * 100
    --     if protectPercent > 100 then
    --         protectPercent = 100
    --     end
    --     self._shadowPercent = protectPercent
    --     self._protect:setPercent(protectPercent)
    --     local hpPercent = hp / self._maxHp * 100
    --     if hpPercent > 100 then
    --         hpPercent = 100
    --     end
    --     self._hp:setPercent(hpPercent)
    -- end
end

--更新底下的血条
function BillBoard:updateHpShadow(needScaleMoving)
    if needScaleMoving then
        local action1 = cc.FadeOut:create(0.2)
        local action2 =
            cc.CallFunc:create(
            function()
                self._hpShadow:setPercent(self._shadowPercent)
                self._hpShadow:setOpacity(255)
            end
        )
        local action = cc.Sequence:create(action1, action2)
        self._hpShadow:runAction(action)
    else
        self._hpShadow:setPercent(self._shadowPercent)
    end
end

--update
function BillBoard:update(f)
end

function BillBoard:showDead()
    local action1 = cc.FadeOut:create(0.2)
    local action2 = cc.RemoveSelf:create()
    local action = cc.Sequence:create(action1, action2)
    self:runAction(action)
end

function BillBoard:updateAnger(count)
    for i = 1, 4 do
        if i <= count then
            self._imageAnger[i]:setVisible(true)
        else
            self._imageAnger[i]:setVisible(false)
        end
    end

    if count > 4 then
        self._angerLabel:setString(count)
        self._angerCountBG:setVisible(true)
    else
        self._angerCountBG:setVisible(false)
    end

    if count >= 4 then
        self:_playAngerEffect()
    else
        self:_hideAngerEffect()
    end
end

function BillBoard:updateBuff(bufflist, camp)
    for i = 1, 4 do
        self._buffIcons[i]:setVisible(false)
    end
    local buffPos = 1
    if #bufflist > 4 then
        for i = #bufflist - 3, #bufflist do
            local data = HeroSkillEffect.get(bufflist[i].configId)
            if data.buff_icon ~= "" then
                local picName = Path.getBuffFightIcon(data.buff_icon)
                self._buffIcons[buffPos]:loadTexture(picName)
                self._buffIcons[buffPos]:setVisible(true)
                buffPos = buffPos + 1
            end
        end
    else
        for i = 1, #bufflist do
            local data = HeroSkillEffect.get(bufflist[i].configId)
            assert(data, "wrong skill id " .. bufflist[i].configId)
            if data.buff_icon ~= "" then
                local picName = Path.getBuffFightIcon(data.buff_icon)
                self._buffIcons[buffPos]:loadTexture(picName)
                self._buffIcons[buffPos]:setVisible(true)
                buffPos = buffPos + 1
            end
        end
    end
end

function BillBoard:_playAngerEffect()
    if not self._angerEffect then
        local EffectGfxNode = require("app.effect.EffectGfxNode")
        local function effectFunction(effect)
            if effect == "effect_nuqi_huoyan_1" or effect == "effect_nuqi_huoyan_4" then
                local subEffect = EffectGfxNode.new("effect_nuqi_huoyan")
                subEffect:play()
                return subEffect
            elseif string.find(effect, "effect_") then
                local subEffect = EffectGfxNode.new(effect)
                subEffect:play()
                return subEffect
            end
        end
        self._angerEffect = G_EffectGfxMgr:createPlayMovingGfx(self, "moving_nuqi", effectFunction, nil, false)
    end
    self._angerEffect:setVisible(true)
end

function BillBoard:_hideAngerEffect()
    if self._angerEffect then
        self._angerEffect:setVisible(false)
    end
end

return BillBoard

File: study/ts/script/fight/views/BillBoard.ts
import { ConfigNameConst } from "../../const/ConfigNameConst";
import { G_ConfigLoader, G_EffectGfxMgr } from "../../init";
import { Lang } from "../../lang/Lang";
import { Color } from "../../utils/Color";
import { Path } from "../../utils/Path";
import UIHelper from "../../utils/UIHelper";
import { FightConfig } from "../FightConfig";
import EffectActor from "./EffectActor";

const { ccclass, property } = cc._decorator;

@ccclass
export default class BillBoard extends cc.Component {

    private _hpShadow: cc.Node;
    private _hp: cc.Node;
    private _imageAnger: cc.Node[];
    private _angerCountBG: cc.Node;
    private _angerLabel: cc.Label;
    private _buffIcons: cc.Sprite[];

    private _needUpdate: boolean;
    private _hpPercent: number;
    private _angerEffect: cc.Node;
    _panelHpBase: cc.Node;
    _protect: any[];
    _maxHp: any;
    _shadowPercent: number;


    public init(name: string, quality: number, rankLevel: number, isPlayer: boolean,
        officelLevel: number, officalInfoName: string, showMark: any[], camp: number, maxHp, trueQuality) {

        let panelHpBase = UIHelper.newSprite(Path.getBattleRes('img_battle_hpbg')).node;
        panelHpBase.name = "panelHpBase"
        this.node.addChild(panelHpBase);
        panelHpBase.setPosition(0, 5);
        this._panelHpBase = panelHpBase;

        this._hpShadow = UIHelper.newSprite(Path.getBattleRes('img_battle_hpshadow')).node;
        this._hpShadow.name = "_hpShadow";
        panelHpBase.addChild(this._hpShadow);
        this._hpShadow.setAnchorPoint(0, 0.5);
        this._hpShadow.setPosition(-40, 0);

        this._protect = [];

        this._hp = UIHelper.newSprite(Path.getBattleRes('img_battle_hp')).node;
        this._hp.name = "_hp";
        panelHpBase.addChild(this._hp);
        this._hp.setAnchorPoint(0, 0.5);
        this._hp.setPosition(-40, 0);

        var angerBase = UIHelper.newSprite(Path.getBattleRes('img_battle_bar_angebg')).node;
        angerBase.name = "_angerBase"
        this.node.addChild(angerBase);
        angerBase.setPosition(0, -2);
        this._imageAnger = [];
        for (var i = 1; i <= 4; i++) {
            let imageAnger = UIHelper.newSprite(Path.getBattleRes('img_battle_bar_ange0' + i)).node;
            imageAnger.name = "_imageAnger" + i;
            imageAnger.setPosition(-25 + (i - 1) * 16, 0);
            this._imageAnger.push(imageAnger);
            angerBase.addChild(imageAnger);
        }
        this._angerCountBG = UIHelper.newSprite(Path.getBattleRes('img_battle_angenumbg')).node;
        this._angerCountBG.name = "_angerCountBG";
        this.node.addChild(this._angerCountBG);
        this._angerCountBG.setPosition(50, 3);
        // this._angerLabel = UIHelper.createWithCharMap(Path.getBattleFont('img_battle_angenum'));
        this._angerLabel = UIHelper.createWithCharMap(Path.getBattleFontLableAtlas('img_battle_angenum'));
        this._angerLabel.fontSize = 18;
        this._angerLabel.name = "_angerLabel";
        this._angerCountBG.addChild(this._angerLabel.node);
        this._angerLabel.string = ('9');
        this._angerLabel.node.setAnchorPoint(0, 0.5);
        // this._angerLabel.node.setPosition(20, 10);
        this._angerLabel.node.setPosition(0, 10);
        if (rankLevel != 0) {
            if (trueQuality == 7 && !isPlayer) {
                name = name + (' ' + (Lang.get('goldenhero_train_text') + rankLevel));
            } else {
                name = name + ('+' + rankLevel);
            }
        }
        var labelName = UIHelper.createWithTTF(name, Path.getCommonFont(), 22);
        labelName.node.name = "_labelName";
        this.node.addChild(labelName.node);
        labelName.node.setPosition(0, 24);
        if (isPlayer) {
            labelName.node.color = (Color.getOfficialColor(officelLevel));
            UIHelper.enableOutline(labelName, Color.getOfficialColorOutline(officelLevel), 2)
        } else {
            labelName.node.color = (Color.getColor(quality));
            UIHelper.enableOutline(labelName, Color.getColorOutline(quality), 2);
        }
        if (officelLevel && officelLevel != 0 && isPlayer) {
            var labelOfficalName = UIHelper.createWithTTF('[' + (officalInfoName + ']'), Path.getCommonFont(), 22);
            this.node.addChild(labelOfficalName.node);
            labelOfficalName.node.setPosition(0, 45);
            labelOfficalName.node.color = (Color.getOfficialColor(officelLevel));
            UIHelper.enableOutline(labelOfficalName, Color.getOfficialColorOutline(officelLevel), 2);
        }
        this._createMarks(showMark, camp);
        this._buffIcons = [];
        for (var i = 1; i <= 4; i++) {
            var params = {};
            let buffIcon = UIHelper.createImage(params);
            buffIcon.name = "_buffIcon" + i;
            this._buffIcons.push(buffIcon.getComponent(cc.Sprite));
            buffIcon.setAnchorPoint(0, 0);
            buffIcon.setPosition(-40 + 20 * (i - 1), -28);
            buffIcon.setScale(0.5);
            this.node.addChild(buffIcon);
            buffIcon.active = (false);
        }
        this._needUpdate = false;
        this._maxHp = maxHp;
        this._hpPercent = 1;
        this._shadowPercent = 1;
        this._angerEffect = null;
        this.node.active = (false);
    }

    _createNewProtect() {
        var protect = UIHelper.newSprite(Path.getBattleRes('img_battle_es')).node;
        this._panelHpBase.addChild(protect);
        protect.setAnchorPoint(cc.v2(0, 0));
        protect.setPosition(cc.v2(1, 2));
        protect.scaleX = (0);
        this._protect.push(protect);
        protect.zIndex = (this._protect.length);
    }

    private _createMarks(showMark: any[], camp: number) {
        for (let i = 0; i < showMark.length; i++) {
            var v = showMark[i];
            var effect = new cc.Node("_showMark" + (i + 1)).addComponent(EffectActor);
            effect.init(FightConfig.MARK[v - 1]);
            effect.setAction('effect', true);
            this.node.addChild(effect.node);
            effect.node.y = (-15);
            if (camp == 2) {
                effect.node.scaleX = (-1);
            }
        }
    }

    public updateHP(hp, protect?) {
        var hpPercent = hp / this._maxHp;
        if (hpPercent > 100) {
            hpPercent = 100;
        }
        this._shadowPercent = hpPercent;
        this._hp.scaleX = (hpPercent);
        if (protect != 0) {
            var protectCnt = Math.ceil(protect / this._maxHp);
            if (protectCnt != 0) {
                while (!this._protect[protectCnt -1]) {
                    this._createNewProtect();
                }
                for (var i = 0; i < protectCnt - 1; i++) {
                    this._protect[i].scaleX =  1;
                }
                if (protectCnt < this._protect.length) {
                    for (var i = this._protect.length -1; i > protectCnt + 1; i += -1) {
                        this._protect[i].removeFromParent();
                        this._protect.splice(i, 1);
                    }
                }
                var lastProtect = protect - (protectCnt - 1) * this._maxHp;
                var protectPercent = lastProtect / this._maxHp;
                this._protect[this._protect.length -1].scaleX = (protectPercent);
            }
        } else {
            for (var j in this._protect) {
                var v = this._protect[j];
                v.removeFromParent();
            }
            this._protect = [];
        }
    }

    public updateHpShadow(needScaleMoving?) {
        if (needScaleMoving) {
            var action1 = cc.fadeOut(0.2);
            var action2 = cc.callFunc(function () {
                this._hpShadow.scaleX = (this._shadowPercent);
                this._hpShadow.opacity = (255);
            }.bind(this));
            var action = cc.sequence(action1, action2);
            this._hpShadow.runAction(action);
        } else {
            this._hpShadow.scaleX = (this._shadowPercent);
        }
    }

    public setUpdate(f) {

    }

    public showDead() {
        var action1 = cc.fadeOut(0.2);
        var action2 = cc.destroySelf();
        var action = cc.sequence(action1, action2);
        this.node.runAction(action);
    }

    public updateAnger(count) {
        for (var i = 1; i <= 4; i++) {
            if (i <= count) {
                this._imageAnger[i - 1].active = true;
            } else {
                this._imageAnger[i - 1].active = false;
            }
        }
        if (count > 4) {
            this._angerLabel.string = (count);
            this._angerCountBG.active = true;
        } else {
            this._angerCountBG.active = false;
        }
        if (count >= 4) {
            this._playAngerEffect();
        } else {
            this._hideAngerEffect();
        }
    }

    public updateBuff(bufflist: any[], camp?) {
        for (var i = 0; i < 4; i++) {
            this._buffIcons[i].node.active = false;
        }
        var buffPos = 0;
        if (bufflist.length > 4) {
            for (var i = bufflist.length - 4; i < bufflist.length; i++) {
                var data = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(bufflist[i].configId);
                if (data.buff_icon != '') {
                    var picName = Path.getBuffFightIcon(data.buff_icon);
                    UIHelper.loadTexture(this._buffIcons[buffPos], picName);
                    this._buffIcons[buffPos].node.active = true;
                    buffPos = buffPos + 1;
                }
            }
        } else {
            for (var i = 0; i < bufflist.length; i++) {
                var data = G_ConfigLoader.getConfig(ConfigNameConst.HERO_SKILL_EFFECT).get(bufflist[i].configId);
                if (data.buff_icon != '') {
                    var picName = Path.getBuffFightIcon(data.buff_icon);
                    UIHelper.loadTexture(this._buffIcons[buffPos], picName);
                    this._buffIcons[buffPos].node.active = true;
                    buffPos = buffPos + 1;
                }
            }
        }
    }

    private _playAngerEffect() {
        if (!this._angerEffect) {
            this._angerEffect = G_EffectGfxMgr.createPlayMovingGfx(this.node, 'moving_nuqi', null, null, false).node;
        }
        this._angerEffect.active = true;
    }

    private _hideAngerEffect() {
        if (this._angerEffect) {
            this._angerEffect.active = false;
        }
    }
}